swarm/pss/notify/notify.go

```diff
@@ -0,0 +1,276 @@
+}
+
+// a notifier has one sendmux entry for each address space it sends messages to
+type sendBin struct {
+	address  pss.PssAddress
+	symKeyId string
+	count    int
+}
+
+// represents a single notification service
+// only subscription address bins that match the address of a notification client have entries. The threshold sets the amount of bytes each address bin uses.
+// every notification has a topic used for pss transfer of symmetrically encrypted notifications
+// contentFunc is the callback to get initial update data from the notifications service provider
+type notifier struct {
+	bins        []*sendBin
+	topic       pss.Topic
+	threshold   int
+	contentFunc func(string) ([]byte, error)
+}
+
+// Controller is the interface to control, add and remove notification services
+type Controller struct {
+	pss       *pss.Pss
+	notifiers map[string]*notifier
+	handlers  map[string]func(string, []byte) error
+	mu        sync.Mutex
+}
+
+// NewController creates a new Controller object
+func NewController(ps *pss.Pss) *Controller {
+	ctrl := &Controller{
+		pss:       ps,
+		notifiers: make(map[string]*notifier),
+		handlers:  make(map[string]func(string, []byte) error),
+	}
+	ctrl.pss.Register(&controlTopic, ctrl.Handler)
+	return ctrl
+}
+
+// IsActive is used to check if a notification service exists for a specified id string
+// Returns true if exists, false if not
+func (c *Controller) IsActive(name string) bool {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return c.isActive(name)
+}
+
+func (c *Controller) isActive(name string) bool {
+	_, ok := c.notifiers[name]
+	return ok
+}
+
+// Request is used by a client to request notifications from a notification service provider
+// It will create a MsgCodeStart message and send asymmetrically to the provider using its public key and routing address
+// The handler function is a callback that will be called when notifications are recieved
+// Fails if the request pss cannot be sent or if the update message could not be serialized
+func (c *Controller) Request(name string, pubkey *ecdsa.PublicKey, address pss.PssAddress, handler func(string, []byte) error) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	msg := NewMsg(MsgCodeStart, name, c.pss.BaseAddr())
+	c.handlers[name] = handler
+	c.pss.SetPeerPublicKey(pubkey, controlTopic, &address)
+	pubkeyid := common.ToHex(crypto.FromECDSAPub(pubkey))
+	smsg, err := rlp.EncodeToBytes(msg)
+	if err != nil {
+		return fmt.Errorf("message could not be serialized: %v", err)
+	}
+	return c.pss.SendAsym(pubkeyid, controlTopic, smsg)
+}
+
+// NewNotifier is used by a notification service provider to create a new notification service
+// It takes a name as identifier for the resource, a threshold indicating the granularity of the subscription address bin, and a callback for getting the latest update
+// Fails if a notifier already is registered on the name
+func (c *Controller) NewNotifier(name string, threshold int, contentFunc func(string) ([]byte, error)) error {
```

Comment:
I think instead of a contectFunc, you should just pass a channel.
and start a go routine that watches subscription and every time it receives an item it calls Notify on the data.

Reply:
mru should just provide a subscription API and its channel could be  passed to this constructor.
Then mru update service can be persisted together with persisting sendbins of notifiers in future.

Reply:
(it was pending discussion about contentfunc/key message)